Fix EZP-26382: ContentType Draft sets IntegerValueValidator min/max to 0

joaoinacio opened the pull request:
JIRA: https://jira.ez.no/browse/EZP-26382

Problem: When creating a ContentType draft from an existing one (Update), the integer fieldtype converter will incorrectly set the Min/Max validators.
This happens as the `validatorState` flag (`data_int4`) is based on min/max values being different from `null`, when they could be `false`.

The fix is to make sure the `validatorState` flag is toggled from the actual fieldTypeConstraints, not storage data.

TODO:
 - [ ] Apply same fix to Float
 - [ ] Tests?

gggeek commented on eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/IntegerConverter.php:
```diff
@@ -67,16 +67,20 @@ public function toFieldValue(StorageFieldValue $value, FieldValue $fieldValue)
-            $storageDef->dataInt1 = $fieldDef->fieldTypeConstraints->validators[self::FLOAT_VALIDATOR_IDENTIFIER]['minIntegerValue'];
-        }
+        $minIntegerValue = isset($fieldDef->fieldTypeConstraints->validators[self::VALIDATOR_IDENTIFIER]['minIntegerValue']) ?
+            $fieldDef->fieldTypeConstraints->validators[self::VALIDATOR_IDENTIFIER]['minIntegerValue'] : false;
```
q: any reason for using FALSE here instead of NULL ?

gggeek commented on eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/IntegerConverter.php:
```diff
@@ -67,16 +67,20 @@ public function toFieldValue(StorageFieldValue $value, FieldValue $fieldValue)
 
-        if (isset($fieldDef->fieldTypeConstraints->validators[self::FLOAT_VALIDATOR_IDENTIFIER]['maxIntegerValue'])) {
-            $storageDef->dataInt2 = $fieldDef->fieldTypeConstraints->validators[self::FLOAT_VALIDATOR_IDENTIFIER]['maxIntegerValue'];
+        if ($minIntegerValue) {
```
q2: what if there IS a value and it is 0? Not sure this IF is correct...

glye replied:
Yes, 0 is a valid value for both min and max.

glye commented on eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/IntegerConverter.php:
```diff
@@ -121,20 +125,20 @@ public function getIndexColumn()
         $state = 0;
 
-        if ($minValue !== null) {
+        if (!empty($minValue)) {
```
`empty(0)` returns true, but 0 is a valid value. So it should rather use something like `is_int` or `is_numeric`, depending on whether the input may be a numeric string.

joaoinacio replied:
good catch, thanks!

andrerom commented:
when is this false? could it be instead cleaned up to use null as already the case?

joaoinacio commented on eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/IntegerConverter.php:
```diff
@@ -121,20 +125,20 @@ public function getIndexColumn()
-     * @param int|null $minValue Minimum int value, or null if not set
-     * @param int|null $maxValue Maximum int value, or null if not set
+     * @param int|null|false $minValue Minimum int value, or null if not set
+     * @param int|null|false $maxValue Maximum int value, or null if not set
```
Yes, `false` is a possible validator value (in fact, the default for max), see https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/Core/FieldType/Validator/IntegerValueValidator.php

It should be possible to use only null, but could make logic/value checking when using this even less simple.

glye replied:
Ok, then we just need some `, or false if ...` here :)

joaoinacio commented:
Updated: set `null` as default if not provided, updated doc comment for `getStorageDefValidatorState()`

@andrerom: as noted above,  FieldDefinition validator can be `false`, and it's actually the default for max.

glye commented on eZ/Publish/Core/Persistence/Legacy/Content/FieldValue/Converter/IntegerConverter.php:
```diff
@@ -121,20 +121,20 @@ public function getIndexColumn()
-     * @param int|null $minValue Minimum int value, or null if not set
-     * @param int|null $maxValue Maximum int value, or null if not set
+     * @param int|false|null $minValue Minimum int value, false if none, or null if not set
+     * @param int|false|null $maxValue Maximum int value, false if none, or null if not set
```
Sorry, I'm still confused :) You mentioned that the default value is `false`, so when is it `null`? When you edit the content type without giving this input it any value?

joaoinacio replied:
Yes, that would be the case, which brings me to think that always changing storage values will result in wrong data on partial updates (only min or max) so a bad approach...

andrerom commented:
I think false might be wrong @joaoinacio, ref https://github.com/ezsystems/ezpublish-kernel/commit/eb5bbcdac288e3e12bc01e4720a6c7fda406c09b

joaoinacio commented:
@andrerom, @lolautruche : AFAICT an issue is that the Float field type validator has changed (false->null), but the Integer field type hasn't (?)

andrerom commented:
> however backporting this would be a BC break, IMO?

It's a bug fix, bug fixes can in some cases slightly break bc to be able to solve a bug. But if you find a way in the FieldType for backports to transform false to null on input that would be great.

@glye weren't you involved in the Float change from RepoForms side? Any additional insight?

glye commented:
@andrerom I see no such change in RepoForms (or remember any).

Btw, I had a quick look at Integer/Type.php, and it seems to be using `null` already. But perhaps some related classes aren't.

joaoinacio commented:
@glye: there is a mismatch between Ingeger Type and `Validator/IntegerValueValidator.php`.
Closing this PR for alt.
